Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -918,6 +971,113 @@ lemma nonsingular [Nontrivial R] {x y : R} (h : E.equation x y) : E.nonsingular
+It is of $j$-invariant $0$ (see `EllipticCurve.ofJ0_j`). -/
+def ofJ0 [Invertible (3 : R)] : EllipticCurve R :=
+  have := invertibleNeg (3 ^ 3 : R)
+  ⟨WeierstrassCurve.ofJ0 R, unitOfInvertible (-3 ^ 3 : R),
+    by rw [unitOfInvertible_val, WeierstrassCurve.ofJ0_Δ R]; norm_num1⟩
+
+lemma ofJ0_j [Invertible (3 : R)] : (ofJ0 R).j = 0 := by
+  simp only [j, ofJ0, WeierstrassCurve.ofJ0_c₄]
+  ring1
+
+/-- When $2$ is invertible, $Y^2 = X^3 + X$ is an elliptic curve.
+It is of $j$-invariant $1728$ (see `EllipticCurve.ofJ1728_j`). -/
+def ofJ1728 [Invertible (2 : R)] : EllipticCurve R :=
+  have := invertibleNeg (2 ^ 6 : R)
+  ⟨WeierstrassCurve.ofJ1728 R, unitOfInvertible (-2 ^ 6 : R),
+    by rw [unitOfInvertible_val, WeierstrassCurve.ofJ1728_Δ R]; norm_num1⟩
+
+lemma ofJ1728_j [Invertible (2 : R)] : (ofJ1728 R).j = 1728 := by
+  field_simp [j, ofJ1728, WeierstrassCurve.ofJ1728_c₄]
+  have := invertibleNeg (2 ^ 6 : R)
+  have := unitOfInvertible_val (-2 ^ 6 : R)
+  rw [unitOfInvertible_val (-2 ^ 6 : R)] -- strange, rw [this] does not work
+  ring1
+
+variable {R}
+
+/-- When $j$ and $j - 1728$ are both invertible,
+$Y^2 + (j - 1728)XY = X^3 - 36(j - 1728)^3X - (j - 1728)^5$ is an elliptic curve.
+It is of $j$-invariant $j$ (see `EllipticCurve.ofJ'_j`). -/
+def ofJ' (j : R) [Invertible j] [Invertible (j - 1728)] : EllipticCurve R :=
+  have := invertibleMul (j ^ 2) ((j - 1728) ^ 9)
+  ⟨WeierstrassCurve.ofJ j, unitOfInvertible <| j ^ 2 * (j - 1728) ^ 9,
+    (WeierstrassCurve.ofJ_Δ j).symm⟩
+
+lemma ofJ'_j (j : R) [Invertible j] [Invertible (j - 1728)] : (ofJ' j).j = j := by
+  field_simp [EllipticCurve.j, ofJ', WeierstrassCurve.ofJ_c₄]
+  have := invertibleMul (j ^ 2) ((j - 1728) ^ 9)
+  have := unitOfInvertible_val <| j ^ 2 * (j - 1728) ^ 9
+  rw [unitOfInvertible_val <| j ^ 2 * (j - 1728) ^ 9]
+  ring1
+
+variable {F : Type u} [Field F] (j : F)
+
+/-- For any element $j$ of a field $F$, there exists an elliptic curve over $F$
+with $j$-invariant equal to $j$ (see `EllipticCurve.ofJ_j`).
+Its coefficients are given explicitly (see `EllipticCurve.ofJ0`, `EllipticCurve.ofJ1728`
+and `EllipticCurve.ofJ'`). -/
+noncomputable def ofJ : EllipticCurve F := by
+  by_cases hj0 : j = 0
+  · by_cases hchar3 : (3 : F) = 0
+    · have : Invertible (2 : F) := by
+        apply invertibleOfNonzero
+        rw [←hchar3]
+        apply_fun (· - 2)
+        norm_num
+        exact zero_ne_one
+      exact ofJ1728 F
+    · have : Invertible (3 : F) := invertibleOfNonzero hchar3
+      exact ofJ0 F
+  · by_cases hj1728 : j = 1728
+    · have : Invertible (2 : F) := by
+        apply invertibleOfNonzero
+        intro h
+        rw [calc (1728 : F) = (2 : F) * (864 : F) := by norm_num, h, zero_mul] at hj1728
+        exact hj0 hj1728
+      exact ofJ1728 F
+    · have : Invertible j := invertibleOfNonzero hj0
+      have : Invertible (j - 1728) := invertibleOfNonzero (Iff.mpr sub_ne_zero hj1728)
+      exact ofJ' j
```

Comment:
> Here is the term mode version using `if then else`, but would require `Classical`: ...

Do you think it's a good idea to just extract several `have`s in my previous proof as private lemmas? In this way the `Classical` is not necessary and the `def` can be written as term mode (while the proof of private lemmas themself still use tactic mode, just like your `two_or_three_ne_zero`).

Reply:
I think your lemma can be rewrite as `(3 : F) = 0 -> (2 : F) ≠ 0`, perhaps in this way the `Classical` is not needed?

Reply:
> In this way the `Classical` is not necessary

No, it does not work 😅 the `if h0 : j = 0 then` itself requires `Classical`, whereas `by_cases hj0 : j = 0` does not require, although they produce the same proof term. Weird.

> `private lemma two_or_three_ne_zero : (2 : F) ≠ 0 ∨ (3 : F) ≠ 0`

More generally, if `R` is a non-trivial ring, `a` and `b` are natural numbers such that
`gcd(a, b) = 1`, then `(a : R) = 0` implies `(b : R) ≠ 0` (otherwise `(1 : R) = 0` contradiction). Do you think this is a useful lemma? Or perhaps it is already in mathlib?

Reply:
Yeah, I think it's a nice lemma to have in mathlib, although I'm not sure where to put it! I thought someone would have written it already but I can't seem to find it anywhere, so you could try to make a quick PR and let others decide...